mindmeld/auto_annotator.py

```diff
@@ -835,5 +901,320 @@ def valid_entity_check(self, entity):
+        do not have an "unknown" unit type.
+        """
+        return [
+            candidate
+            for candidate in candidates
+            if not (
+                candidate["dim"] == "amount-of-money"
+                and candidate["value"].get("unit") == "unknown"
+            )
+        ]
+
+
+class TranslationDucklingAnnotator(Annotator):
+    """Custom Annotator class used to generate annotations."""
+
+    def __init__(self, *args, **kwargs):
+        """Initializes a TranslationDucklingAnnotator.
+
+        Args:
+            app_path (str): The location of the MindMeld app.
+            annotation_rules (list): List of Annotation rules.
+            en_annotator (SpacyAnnotator): A Spacy Annotator with language set to English ("en").
+            translator (str): A translator to use such as 'GoogleTranslator' or 'NoOpTranslator'.
+            language (str, optional): Language as specified using a 639-1/2 code.
+            locale (str, optional): The locale representing the ISO 639-1 language code and \
+                ISO3166 alpha 2 country code separated by an underscore character.
+            overwrite (bool): Whether to overwrite existing annotations with conflicting spans.
+            unannotate_supported_entities_only (bool): Only allow removal of supported entities.
+            unannotation_rules (list): List of Annotation rules.
+        """
+        super().__init__(*args, **kwargs)
+        assert (
+            self.language != ENGLISH_LANGUAGE_CODE
+        ), "The 'language' for a TranslationDucklingAnnotator cannot be set to English."
+        translator = kwargs.get("translator")
+        self.translator = TranslatorFactory().get_translator(translator)
+        self.en_annotator = kwargs.get("en_annotator") or SpacyAnnotator(
+            app_path=self.app_path,
+            language=ENGLISH_LANGUAGE_CODE,
+            locale=ENGLISH_US_LOCALE,
+        )
+
+    def parse(self, sentence, entity_types=None, **kwargs):
+        """
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                    possible entity types will be parsed.
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
+            sentence,
+            entity_types=entity_types,
+            language=self.language,
+            locale=self.locale,
+        )
+        en_sentence = self.translator.translate(  # pylint: disable=E1128
+            sentence, target_language=ENGLISH_LANGUAGE_CODE
+        )
+        en_entities = self.en_annotator.parse(en_sentence, entity_types=entity_types)
+        final_candidates = []
+        for entity in en_entities:
+            value_matched_candidates = []
+            for candidate in candidates:
+                if entity["dim"] != candidate["entity_type"]:
```

Comment:
I would recommend adding comments for these rules to future readability. Its highly specific.

Reply:
Added